contracts/proposal/dao-proposal-condorcet/src/contract.rs

```diff
@@ -0,0 +1,279 @@
+//
+// that being true, you will never be able to create a proposal that
+// can not be voted on and executed inside gas limits.
+//
+// in terms of storage costs:
+//
+// propose: proposal_load + proposal_store + tally_load + tally_store + config_load
+// execute: proposal_load + proposal_store + tally_load
+// vote:                                     tally_load + tally_store               + vote_load + vote_store
+//
+// so we are good so long as:
+//
+// `vote_load + vote_store <= proposal_load + proposal_store + config_load`
+//
+// this is true so long as a vote is smaller than a proposal in
+// storage which is true because proposals store `choices =
+// Vec<Vec<CosmosMsg>>`, `choices.len() = vote.len()`, vote is a
+// `Vec<u32>`, even an empty vec must contain it's length which is a
+// usize, so `sizeof(Vec<u32>) <= sizeof(Vec<usize>) <=
+// sizeof(Vec<Vec<CosmosMsg>) => sizeof(vote) <= sizeof(proposal)`.
+//
+// in terms of other costs:
+//
+// propose: query_voting_power + compute_winner [2]
+// execute: query_voting_power
+// vote:    query_voting_power + compute_winner
+//
+// so we're good there as well.
+//
+// [1] we need to be gas efficent in this way because the size of the
+//     Tally type grows with candidates^2 and thus can be too large to
+//     load from storage. we need to make sure that if this is the
+//     case, the proposal fails to be created. the bad outcome we're
+//     trying to avoid here is a proposal that is created but can not
+//     be voted on or executed.
+// [2] Tally::new computes the winner over the new matrix so that this
+//     is the case.
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Propose { choices } => execute_propose(deps, env, info, choices),
+        ExecuteMsg::Vote { proposal_id, vote } => execute_vote(deps, env, info, proposal_id, vote),
+        ExecuteMsg::Execute { proposal_id } => execute_execute(deps, env, info, proposal_id),
+        ExecuteMsg::Close { proposal_id } => execute_close(deps, env, info, proposal_id),
+
+        ExecuteMsg::SetConfig(config) => execute_set_config(deps, info, config),
+    }
+}
+
+fn execute_propose(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    choices: Vec<Choice>,
+) -> Result<Response, ContractError> {
+    let dao = DAO.load(deps.storage)?;
+    let sender_voting_power = get_voting_power(deps.as_ref(), info.sender.clone(), &dao, None)?;
+    if sender_voting_power.is_zero() {
+        return Err(ContractError::ZeroVotingPower {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+
+    let id = next_proposal_id(deps.storage)?;
+    let total_power = get_total_power(deps.as_ref(), &dao, None)?;
+
+    if choices.is_empty() {
+        return Err(ContractError::ZeroChoices {});
+    }
+
+    let none_of_the_above = Choice { msgs: vec![] };
```

Comment:
we'd also want to check if there are at least two choices, and that choice length is below some bound

Reply:
i think only one choice is actually fine (`!is_empty()` above) because in that case it becomes exactly the same as a `dao-proposal-single` election. down the line we might even want to replace proposal-single with this.